build/root.build

```diff
@@ -8,3 +8,9 @@ cxx{*}: extension = cc
 # The test target for cross-testing (running tests under Wine, etc).
 #
 test.target = $cxx.target
+
+# Disable support for building shared libs on windows with msvc
+# Upstream does not support this
+# Public headers are not marked appropriately with declspec(dllimport) and declspec(dllexport)
+# Causes severe problems with linking due to static variables and inline functions in headers
+assert! ($config.bin.lib != 'static' && $cxx.target.system == 'win32-msvc') 'Google Benchmark does not support building shared libs with MSVC. Explicitly set config.bin.lib=static when using MSVC'
```

Comment:
Correct me if I'm wrong, but won't this force dependent projects to also be built as `static`? Or can one specify to only build `google-benchmark` as static?

`bdep init config.bin.lib=static` still prints this error though (adding a `print` shows this value twice, and it's `both` first time, `static` second). It also fails when doing `bdep deinit` (since it appears to check this then as well). Anyways, I don't think this sort of hard enforcement is the correct way to go about this (can't find it now but I remember reading it somewhere in the build2 docs/issues). Also, as mentioned [here](https://github.com/build2-packaging/google-benchmark/issues/3#issuecomment-1006137466), it worked fine before by simply importing the static target (`google-benchmark%liba{benchmark}`).

Then again, it would be confusing to build `both` and only get the static `benchmark` target...

Reply:
As I understand all `build2` configs are amalgamations. And during `init`, one can pass package specific options.

Importing `liba{}` vs `libs{}` vs `lib{}` is consumer’s choice unless there is very special underlying reason. Consumers typically would want to use `lib{}` unless there `liba{}` only export for example `liba{benchmark-main}`.

So options are:
* Support `liba{}` only, for all platforms and compilers because one specific combination doesn’t support shared libs.
* Disable build on the single combination and ask the user to pass the `config` option.

I would like to support `libs{}` with gcc and clang and mingw, so can’t enforce building only `liba{benchmark}`.

The config option can be for the whole amalgamation or for the this specific package. This is again the consumers choice. But the general idea is that this choice only needs be made for only one specific platform and compiler combination.

Reply:
> bdep init config.bin.lib=static still prints this error though (adding a print shows this value twice, and it's both first time, static second).

On a de-initialised configuration could you try the following command
```
bdep init { config.bin.lib=static }+ "?google-benchmark"
```
The extra `{ }+` is for argument grouping to apply only to `google-benchmark` while using default configuration for all other packages.

Reply:
> As I understand all `build2` configs are amalgamations. And during `init`, **one can pass package specific options**.

How? I can't find anything about this in the documentation. I'd expect something like `config.benchmark.bin.lib=static`. And still, `config.bin.lib=static` didn't work either.

> Importing `liba{}` vs `libs{}` vs `lib{}` is consumer’s choice unless there is very special underlying reason. Consumers typically would want to use `lib{}` unless there `liba{}` only export for example `liba{benchmark-main}`.
> I would like to support `libs{}` with gcc and clang and mingw, so can’t enforce building only `liba{benchmark}`.
> 

I agree that `lib{}` should be used for other platforms. I rather thought something like `if msvc -> setup liba{benchmark} only` (so you get an error about missing `benchmark` target if using `lib{benchmark}` and building shared, or explicitly `libs{benchmark}`).

Wouldn't it be better to not have any enforcement at all, perhaps only a warning referring to [this](https://github.com/google/benchmark/issues/639)?

Or... Implement your own version of [`WINDOWS_EXPORT_ALL_SYMBOLS`](https://cmake.org/cmake/help/latest/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html) (better yet if it was available in `build2`). :)

Reply:
> On a de-initialised configuration could you try the following command
> 
> ```
> bdep init { config.bin.lib=static }+ "?google-benchmark"
> ```

Yeah... cmd nor powershell is swallowing that command. :)

Reply:
> Or... Implement your own version of WINDOWS_EXPORT_ALL_SYMBOLS (better yet if it was available in build2). :)

Reply:
> Or... Implement your own version of WINDOWS_EXPORT_ALL_SYMBOLS (better yet if it was available in build2). :)

`build2` actually has this. Boris implemented this feature and is available starting 0.14.0.
As I mentioned in my comment, building a shared lib works fine. The under laying problem is not the export of the symbols. Rather, it's import. One commit prior in this PR, with ref: fcc669e, I already implemented symbol exports using `build2`'s `def{}` which works well and exports all symbols.

Reply:
> Wouldn't it be better to not have any enforcement at all, perhaps only a warning referring to this?

I went with the more conservative approach first.
It is better not to initiate the build at all when the library configuration is not as expected. An upfront `assert` can save quite a lot of time and effort rather than allowing consumers to mistakenly link against the shared lib.

To make it work properly, consumers would have to pay attention to a single log message in-between a wall of text during the build. In the end, depending on how individual benchmarks are written, they can have very weird linker issue, like the one I had. By default, `config.bin.exe.lib=shared static`, which means a shared lib is preferred to a static lib. So consumers may not even realise the underlaying problem. `link.exe` would happily create an executable and devs would redo the same investigation all over again when something goes wrong.

Reply:
> I think the problem is with escaping certain characters. I personally using git-bash so I got an error for only the `?google-benchmark` part. On `cmd` the same command worked for me. On `powershell` quotes around `{ config.bin.lib=static }+` worked:
> 
> ```
> bdep init "{ config.bin.lib=static }+" "?google-benchmark"
> ```

Not seeing the same thing here... I've tried all sorts of combinations, and it fails in all but one case with bash (see below).

**Powershell:**
Does not work:
```
bdep init "{ config.bin.lib=static }+" "?google-benchmark"
error: missing value for option '-n'
```
Does not work:
```
bdep init -C -@msvc cc config.cxx=cl.exe "{ config.bin.lib=static }+" "?google-benchmark"
error: expected variable assignment instead of '{ config.bin.lib=static }+'
  info: use double '--' to treat this argument as buildspec
```
**Bash:**
Works:
```
bdep init @msvc-bash { config.bin.lib=static }+ "?google-benchmark"
```
Does not work: 
```
bdep init @msvc-bash "{ config.bin.lib=static }+" "?google-benchmark"
```
Does not work:
```
bdep init -C @msvc-bash cc config.cxx=cl.exe { config.bin.lib=static }+ "?google-benchmark"
error: unexpected grouped argument 'config.bin.lib=static' for argument '?google-benchmark'
```

So only one case worked, and that requires me to first create the config (which will then error out on the assert), then deinit and init again.

> `config.build` can even be hand edited.

Again, can't achieve this. Adding this to `config.build` (quotes, no quotes, whatever):
`{ config.bin.lib=static }+ ?google-benchmark`
Gives:
```
bdep init -C -@msvc cc config.cxx=cl.exe
error: Google Benchmark does not support building shared libs with MSVC. Explicitly set config.bin.lib=static when using MSVC
```

> I think the same is found at: [bdep-doc](https://build2.org/bdep/doc/bdep.xhtml) and bdep-argument-grouping.

Thanks!

So TBH, given all the hassle surrounding this (unless you find my error in how I setup `config.build` incorrectly), I'd vote for this:

Re-add `def{...}` & remove the one test that uses `PerfCounters::kSupported` & fails. I've tested it and all just works. No errors, no special configs for a certain platform etc. See attached file for my draft
[Export-all-symbols-when-building-shared-for-windows.txt](https://github.com/build2-packaging/google-benchmark/files/7842197/Export-all-symbols-when-building-shared-for-windows.txt)
.

Reply:
@helmesjo,
I'm unable to reproduce these errors.
At my end the commands work with various quoting on `git-bash`, `cmd`, `powershell`. Tried `bash` and `zsh` on linux as well.
In principle, there is a way to do it. The commands are documented. If you are facing issues with `bdep init` then it's probably a good idea to open an issue in the [build2/bdep](https://github.com/build2/bdep) repo. The initialisation syntax itself has nothing to do with this package. They are orthogonal.

Reply:
> So only one case worked, and that requires me to first create the config (which will then error out on the assert), then deinit and init again.

This also what I experienced so far. One can add a feature request to allow both config creation + config passing in [build2/bdep](https://github.com/build2/bdep). Again this is not related to the package. The package only requires a certain configuration. How this configuration is achieved, like using 2 `bdep` commands instead of 1, or using `b config` form or using `bpkg config` is not relevant.

> Again, can't achieve this. Adding this to config.build (quotes, no quotes, whatever):
{ config.bin.lib=static }+ ?google-benchmark

This syntax should not be used when editing `config.build`. Please have a look at [build2 docs. Section 1.2](https://build2.org/build2/doc/build2-build-system-manual.xhtml#intro-proj-struct) for a general overview and project/configuration structure. For details about the `config` module have a look at [build2 docs. Section 1.4.1](https://build2.org/build2/doc/build2-build-system-manual.xhtml#intro-operations-config)
In summary, please place the variable `config.bin.lib=static` under `<your-configuration-path>/google-benchmark-<version-number>/build/config.build`.
Notice that, the configuration itself has a minimum of 3 `config.build` files.
* 1 under `<your-configuration-path>` -> This is a global configuration file applied to all packages within the config
* 1 under `<your-main-pkg>` -> Pkg specific `config.build`
* 1 under `google-benchmark-<versionnumber>` -> `google-benchmark` specific `config.build`

`config.build` can be hand made or generated. They are meant to be hand-edited post generation as well.
Again, configuration setup is not directly related to this package and is off topic.
I found this [video](https://youtu.be/hHRaZy1LZPA) to be quite useful in getting started with `build2`.

Reply:
> Re-add def{...} & remove the one test that uses PerfCounters::kSupported (which is under the internals namespace anyways) & fails. I've tested it and all just works. No errors, no special configs for a certain platform etc.
See my draft here (just git apply onto this branch): Export-all-symbols-when-building-shared-for-windows.txt

As I said, building a shared lib is no problem. And for simple tests this is ok and works fine. Problem comes when consumers utilise more features offered by the library. By allowing `libs{benchmark}` to build, we create a false sense that everything works, when we already know it's not the right thing to do. This is not a good design practice in general for both `c++` code and `buildfile`s. Flagging an error and asking the user to correct an incorrect/unsupported configuration, in my opinion is a fair way to deal with it. This is not limited to a set of tests either. When the package precondition is violated, it does not make any sense to continue building `libs{benchmark}`. 

My ideal solution would be to disable `libs{benchmark}` with `msvc`, log a warning and continue building only `liba{benchmark}`. At consumer side, access it as `lib{benchmark}` and get `liba{benchmark}` with `msvc`, on all other platforms, get what is set by the user. If the consumer forces `config.bin.lib=shared` then error out under `msvc`. I don't know how this can be achieved currently. @boris-kolpackov any idea/suggestion on how to achieve this?

Reply:
> As I said, building a shared lib is no problem. And for simple tests this is ok and works fine. Problem comes when consumers utilise more features offered by the library. By allowing `libs{benchmark}` to build, we create a false sense that everything works, when we already know it's not the right thing to do. This is not a good design practice in general for both `c++` code and `buildfile`s. Flagging an error and asking the user to correct an incorrect/unsupported configuration, in my opinion is a fair way to deal with it. This is not limited to a set of tests either. When the package precondition is violated, it does not make any sense to continue building `libs{benchmark}`.

My bad, I was under the impression that the only issue there was with building shared libs was one or two `static` variables (and I only saw "internal" ones used solely by tests). Thanks for clarifying.

> My ideal solution would be to disable `libs{benchmark}` with `msvc`, log a warning and continue building only `liba{benchmark}`. At consumer side, access it as `lib{benchmark}` and get `liba{benchmark}` with `msvc`, on all other platforms, get what is set by the user. If the consumer forces `config.bin.lib=shared` then error out under `msvc`.

Isn't this what you describe?
```
# root.build

if ($cxx.target.system == 'win32-msvc')
{
  if ($bin.lib == 'both')
  {
    warn 'Shared libs not supported for Windows, building static only.'
    bin.lib=static
  }
  elif($bin.lib == 'shared')
  {
    fail 'Shared libs not supported for Windows'
  }
}
```

Reply:
You can achieve this by overriding `bin.lib` in your `build/root.build`. For example, add something along these lines after loading the `cxx` module:

```
if ($cxx.target.class == 'windows')
{
  assert ($bin.lib != 'shared') 'only static variant can be built on Windows'
  bin.lib = static
}
```

For completeness you will also want to disable importing of the shared variant by modifying your `build/export.build` along these lines:

```
$out_root/
{
  include libhello/
}

if ($($out_root/libhello/ cxx.target.class) == 'windows')
  assert ($import.target != libs{hello}) "only static on Windows"

export $out_root/libhello/$import.target
```

I wouldn't recommend issuing a warning (there is nothing the user of your library can do about it), but if you want, you can modify the code in `build/root.build` along these lines:

if ($cxx.target.class == 'windows')
{
  switch $bin.lib
  {
    case 'shared'
      fail 'only static variant can be built on Windows'
    case 'both'
    {
      warn 'only static variant will be built on Windows'
      bin.lib = static
    }
  }
}

Reply:
Cool!! Thanks @helmesjo and @boris-kolpackov, I never tried to set `bin.lib` within the `buildfile`! I’ll try it out and put it in!

Reply:
@boris-kolpackov 
Is there anything additional that can be done to error out during importation against `google-benchmark%libs{benchmark}` (notice `libs{}`) when building `config.bin.lib=both`? Right now it will go through (even though there is no shared target) & instead fail during linking.